pkg/validation/runner.go

```diff
@@ -0,0 +1,143 @@
+package validation
+
+import (
+	"context"
+	"reflect"
+	"sync"
+
+	"github.com/aws/eks-anywhere/pkg/errors"
+)
+
+// Validatable is anything that can be validated.
+type Validatable[O any] interface {
+	Copiable[O]
+}
+
+// Copiable is anything that can be copied.
+type Copiable[O any] interface {
+	DeepCopy() O
+}
+
+// Validation is the logic for a validation of a type O.
+type Validation[O Validatable[O]] func(ctx context.Context, obj O) error
+
+// Runner allows to compose and run validations.
+type Runner[O Validatable[O]] struct {
+	validations []Validation[O]
+	config      *RunnerConfig
+}
+
+// RunnerConfig contains the configuration for a Runner.
+type RunnerConfig struct {
+	maxJobs int
+}
+
+// RunnerOpt allows to configure a Runner with optional parameters.
+type RunnerOpt func(*RunnerConfig)
+
+// WithMaxJobs sets the maximun number of concurrent routines the runner will use.
+func WithMaxJobs(m int) RunnerOpt {
+	return func(c *RunnerConfig) {
+		c.maxJobs = m
+	}
+}
+
+// NewRunner constructs a new Runner.
+func NewRunner[O Validatable[O]](opts ...RunnerOpt) *Runner[O] {
+	r := &Runner[O]{
+		config: &RunnerConfig{
+			maxJobs: 8,
+		},
+	}
+
+	for _, opt := range opts {
+		opt(r.config)
+	}
+
+	return r
+}
+
+// Register adds validations to the Runner.
+func (r *Runner[O]) Register(validations ...Validation[O]) {
+	r.validations = append(r.validations, validations...)
+}
+
+// RunAll runs all validations concurrently and waits until they all finish,
+// aggregating the errors if present. obj must not be modified. If it is, this
+// indicates a programming error and the method will panic.
+func (r *Runner[O]) RunAll(ctx context.Context, obj O) errors.Aggregate {
```

Comment:
Good question. I was thinking originally that we could possibly offer a `Run` that would just return errors through a channel, so the caller can handle them as they come as opposed to block and wait.

But tbh, since I'm not sure we'll ever do that, I'm fine changing it to `Run` if you prefer it.